Don't hold the GIL during calls

iamsrp opened the pull request:
```
Release the GIL during calls to the C library. This allows other threads in the
Python process to progress while the library is busy talking etc.
```

This commit means that pyfestival can be invoked in worker thread while the main thread carries on doing whatever it is that it does.

iamsrp

iamsrp committed:
Don't hold the GIL during calls

Release the GIL during calls to the C library. This allows other threads in the
Python process to progress while the library is busy talking etc.

Since the Festival library is not guaranteed to be thread-safe, we now use a
mutex to guard calls into it, thus ensuring that only one thrad at a time may
use the library. Previously the GIL was effectively doing this for us.

pjrobertson commented:
Code looks good to me. @iamsrp do these changes only have an effect when pyfestival is called from a non-main thread? what happens when called on the main thread?

Thanks for the input @iamsrp

iamsrp commented:
This change is thread-agnostic; it will have the same behaviour wherever it's called. 

However, festival doesn't look very threadsafe itself (I think that this is down to the festival libraries, nothing to do with the Python wrapper layer). If you import pyfestival in one thread and attempt to use it in another then you get mysterious errors out of it. It's fine to import it in a thread and use it directly there but you can't have two active instances at once in different threads. (It's fine for the thread to die and to then import and use it again in a new thread.)

Peering at the festival code (at a very cursory level) I don't think that the authors intended any thread safety so it's impressive that it works at all without core dumping or similar.

Oh, and for what it's worth, I did chuckle that all the pyfestival commits were in Welsh.

Cheers,

iamsrp

pjrobertson commented:
> Peering at the festival code (at a very cursory level) I don't think that the authors intended any thread safety
> 
That’s what I assumed, and makes me a little bit wary of us making these changes here. I’d call this feature ‘experimental’. In which case, perhaps we should have the option to enable/disable threads at build time, with it set to ‘disabled’ by default. Can you please go ahead and alter your commit to add a pre-processor directive  `#define ENABLE_THREADS false/true` which will allow threads to be enabled/disabled at build time. That’s something I think that’s best.

> Oh, and for what it's worth, I did chuckle that all the pyfestival commits were in Welsh.
> 
Right! Hope you were able to follow them! ;-)


> On 13 Nov 2020, at 14:22, iamsrp <notifications@github.com> wrote:
> 
> 
> This change is thread-agnostic; it will have the same behaviour wherever it's called.
> 
> However, festival doesn't look very threadsafe itself (I think that this is down to the festival libraries, nothing to do with the Python wrapper layer). If you import pyfestival in one thread and attempt to use it in another then you get mysterious errors out of it. It's fine to import it in a thread and use it directly there but you can't have two active instances at once in different threads. (It's fine for the thread to die and to then import and use it again in a new thread.)
> 
> Peering at the festival code (at a very cursory level) I don't think that the authors intended any thread safety so it's impressive that it works at all without core dumping or similar.
> 
> Oh, and for what it's worth, I did chuckle that all the pyfestival commits were in Welsh.
> 
> Cheers,
> 
> iamsrp
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/techiaith/pyfestival/pull/11#issuecomment-726548351>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABEXHZCMCB7OFOD2G6NPOLSPTGBXANCNFSM4TOZ7QGQ>.
>

iamsrp commented:
That's a fair criticism. Just because it works today it might not work tomorrow and we don't want to be at the mercy of some change in festival breaking us here. 

However, looking at the problem I don't think that selectively enabling this feature is a long term solution (for the reason stated above). The basic problem with my commit is that I've removed lock protection by virtue of disabling the GIL during the duration of the calls. We still need some form of lock to ensure that calls the festival are effectively single-threaded / synchronous. As such, I think that holding a local mutex for the duration of the calls is the right thing to do here; effectively replacing one lock with another. Provided that this mutex is acquired after we've dropped the GIL (and dropped before we re-acquire the GIL) we should not encounter deadlock situations. I'll also use it to protect what the GIL was effectively protecting elsewhere (e.g. the file operations).

Adding the mutex is easy here; I'll look to do that soon and then to create a new pull request with the changes. I'll close this one out at the same time that I do that.

Cool?

iamsrp

iamsrp commented:
Well, I made the changes in my fork but my github kung fu is sorely lacking. I did the naive thing in my tree and reverted the original commit so as to now apply the new version. It looks like the PR has picked up these new changes, I guess because it's all in the same branch. 

Anyways, the new version does its own locking to handle where we're not using the GIL. I tested it and it seems fine and happy. Not sure how you'll feel about the formatting of the code though; I kinda guessed what might be reasonable but feel free to edit to your preferred style.

iamsrp

pjrobertson commented:
Alright, this looks good to me! Can you squash all the changes down into a single commit and force push, then I can merge directly into master.

iamsrp commented:
Got it; I didn't realise that one could use `--force` to rewrite history in the main repo. In theory it's done now, and it _looks_ like the PR has been updated accordingly. Let me know if you think otherwise.

Cheers,

iamsrp

iamsrp commented:
PS Aside from polling, is there a way for me to know when the changes will have made it into the pip repo? I ask simply so that I can update the guidance etc. in my [use-case code](https://github.com/iamsrp/dexter) accordingly.

iamsrp committed:
Avoid calling sayText with empty strings

If you pass a NULL or an empty string to sayText() then it will result in a SEGV. Guard against this by not doing that.

iamsrp committed:
Added a note about threading

The underlying library is not threadsafe. This applies to Python usage too.

iamsrp commented:
I also just fixed a SEGV which gets thrown if you do `festvial.sayText('')`. Updated the README too.

Since these are separate changes I kept them as separate commits.

pjrobertson commented:
Hey @iamsrp  - sorry for not getting to this sooner. Latest commits look good. Merged

RE the pip repo - I'm not sure when we'll be able to get to updating that. I suggest you use this repo for your project (`pip install git+git://github.com/techiaith/pyfestival.git#egg=pyfestival`)

We'll see if we can update the pip repo @DewiBrynJones ?

iamsrp commented:
No worries. I didn't want to get naggy :) We're there now.

The pip repo is more for other folks using my stuff though, since there are probably only about 2 of them in the world right now I doubt that it's the end of the world if it takes a little while to happen.

Cheers,

srp

DewiBrynJones commented:
Diolch yn fawr Patrick! And thank you srp for your fix!

@stefanoghazzali and I can take a look at updating https://pypi.org/project/pyfestival/ 

Of interest to @str20tbl as well.

duwudi commented:
> If you import pyfestival in one thread and attempt to use it in another then you get mysterious errors out of it. It's fine to import it in a thread and use it directly there but you can't have two active instances at once in different threads.

@iamsrp thanks for your work on this! Stumbling across the same problem you mention here, the error is `SIOD ERROR: the currently assigned stack limit has been exceeded`. There's a mention of this problem [here too](https://www.linuxquestions.org/questions/linux-general-1/festival-c-api-and-pthread-839667/#post4136045) and it seems they have a similar fix in that they initialise festival from within the thread. I've tried `importlib.reload` and various other methods to try and get it working but to no avail - any other thoughts on how to fix it?

iamsrp commented:
I've not dug into the Festival code too deeply but a _cursory_ glance gave me the impression that fixing this would require a reasonable amount of work. My sense was that the code wasn't really intended to be used in a threaded environment and it makes assumptions accordingly. (This doesn't feel unreasonable; you kinda only want one voice speaking at once etc.) Again, this was just a sniff-test so I could be completely wrong here, but that was why I figured that working around the problem was the most pragmatic course of action.
